X509 Certificate user authentication

brat000012001 opened the pull request:
Please do not merge, docs & tests are a work in progress.
User authentication via X.509 client certificates. The implementation includes support for browser and direct grant flows. See services\src\main\java\org\keycloak\authentication\authenticators\x509\README.md for a brief summary of proposed changes/additions. 

For OCSP certificate revocation checking to work properly, [WFCORE-1705](https://issues.jboss.org/browse/WFCORE-1705) must be resolved.

brat000012001 committed:
KEYCLOAK-4335: x509 client certificate authentication

Started on implementing cert thumbprint validation as a part of x509 auth flow. Added a prompt screen to give users a choice to either log in based on the identity extracted from X509 cert or to continue with normal browser login flow authentication; clean up some of the comments

x509 authentication for browser and direct grant flows. Implemented certificate to user mapping based on user attribute

Implemented CRL and OCSP certificate revocation checking and added corresponding configuration settings to set up responderURI (OCSP), a location of a file containing X509CRL entries and switiches to enable/disable revocation checking; reworked the certificate validation; removed superflous logging; changed the certificate authentication prompt page to automatically log in the user after 10 seconds if no response from user is received

Support for loading CRL from LDAP directory; finished the CRL checking using the distribution points in the certificate; updated the instructions how to add X509 authentication to keycloak authentication flows; minor styling changes

Stashing x509 unit test related changes; added the steps to configure mutual SSL in WildFly to the summary document

A minor fix to throw a security exception when unable to check cert revocation status using OCSP; continue working on README

Changes to the formating of the readme

Added a list of features to readme

Fixed a potential bug in X509 cert user authenticator that may cause NPE if the client certificate does not define keyusage or extended key usage extensions

Fixed compile time errors in X509 validators caused by the changes to the user credentials model in upstream master

Removed a superfluous file created when merging x509 and main branches

X509 authentication: removed the PKIX path validation as superflous

Reverted changes to the AbstractAttributeMapper introduced during merging of x509 branch into main

Merge the unit tests from x509 branch

added mockito dependency to services project; changes to the x509 authenticators to expose methods in order to support unit tests; added a default ctor to CertificateValidator class to support unit testing; updated the direct grant and browser x509 authenticators to report consistent status messages; unit tests to validate X509 direct grant and browser authenticators; fixed OCSP validation to throw an exception if the certificate chain contains a single certificate; fixed the CRL revocation validation to only use CRL distribution point validation only if configured

CRL and OSCP mock tests using mock netty server. Changed the certificate validator to better support unit testing.

changes to the mockserver dependency to explicitly exclude xercesImpl that was causing SAMLParsingTest to fail

Added a utility class to build v3 certificates with optional extensions to facilitate X509 unit testing; removed supoerfluous certificate date validity check (undertow should be checking the certificate dates during PKIX path validation anyway)

X509: changes to make configuring the user identity extraction simplier for users - new identity sources to map certificate CN and email (E) attributes from X500 subject and issuer names directly rather than using regular expressions to parse them

X509 fixed a compile error caused by the changes to the user model in master

Integration tests to validate X509 client certificate authentication

Minor tweaks to X509 client auth related integration tests

CRLs to support x509 client cert auth integration tests

X509: reverted the changes to testrealm.json and updated the test to configure the realm at runtime

X509 - changes to the testsuite project configuration to specify a path to a trust store used to test x509 direct grant flow; integration tests to validate x509 authentication in browser and direct grant flows; updated the client certificate to extend its validatity dates; x509 integration tests and authenticators have been refactored to use a common configuration class

X509 separated the browser and direct grant x509 authenction integration tests

x509 updated the authenticator provider test to remove no longer supported cert thumbprint authenticator

x509 removed the dependency on mockito

x509 re-implemented OCSP certificate revocation client used to check revocation status when logging in with x509 certificate to work around the dependency on Sun OCSP implementation; integration tests to verify OCSP revocation requests

index.txt.attr is needed by openssl to run a simple OCSP server

x509: minor grammar fixes

Add OCSP stub responder to integration tests

This commit adds OCSP stub responder needed for the integration tests,
and eliminates the need to run external OCSP responder in order to run
the OCSP in X509OCSPResponderTest.

Replace printStackTrece with logging

This commit replaces call to printStackTrace that will end up going to
the stderr with logging statement of WARN severity.

Remove unused imports

Removed unused imports in
org.keycloak.authentication.authenticators.x509 package.

Parameterized Hashtable variable

Removed unused CertificateFactory variable

Declared serialVersionUID for Serializable class

Removed unused CertificateBuilder class

The CertificateBuilder was not used anywhere in the code, removing it to
prevent technical debt.

Removing unused variable declaration

`response` variable is not used in the test, removed it.

Made sure InputStreams are closed

Even though the InputStreams are memory based, added try-with-resources
to make sure that they are closed.

Removed deprecated usage of URLEncoder

Replaced invocation of deprecated method from URLEncoder with Encode
from Keycloak util package.

Made it more clear how to control OCSP stub responder in the tests

X509 Certificate user authentication: moved the integration unit tests into their own directory to fix a failing travis test job

KEYCLOAK-4335: reduced the logging level; added the instructions how to run X.509 related tests to HOW-TO-RUN.md doc; removed README.md from x509 folder; removed no longer used ocsp profile and fixed the exclusion filter; refactored the x509 base test class that was broken by the recent changes to the integration tests

KEYCLOAK-4335: fixed a few issues after rebasing

brat000012001 commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java:
```diff
@@ -0,0 +1,726 @@
+    }
+
+    private static void validateKeyUsage(X509Certificate[] certs, int expected) throws GeneralSecurityException {
+        boolean[] keyUsageBits = certs[0].getKeyUsage();
```
X509Certificate.getKeyUsage may return null.

stianst commented:
Thanks for the PR, we are trying to find time to review it on our end. What is the status in your opinion? Is it completed or waiting review before you'll complete it?

brat000012001 commented:
I as far the requirements go, it is completed. The biggest showstopper for me right now are the tests (Selenium/HTTPS, CLR DP revocation testing, OCSP revocation testing). I am certainly willing to discuss and augment the implementation if there are any impeding issues, conceptual or otherwise, found during the review process.

stianst commented:
I've not had a chance to look at it in detail, but will try next week once we've released 2.2. Had a quick scan and it looks OK, one thing that I noticed is that you're not using the Truststore SPI we added a while back. The intent is that we have a single place to configure the truststore and that it would be possible to implement other types of providers rather than just a truststure file. In the future we'd like to add an option to manage trusted certificates through the admin console.

The provider interface is https://github.com/keycloak/keycloak/blob/master/server-spi/src/main/java/org/keycloak/truststore/TruststoreProvider.java.

Does that work for you? If not why not and can we change it somehow so it does?

stianst commented:
BTW we'd also like to make Undertow eventually use the same truststore as well, but that's not possible until WildFly 11 as it depends on Elytron.

brat000012001 commented:
Ideally, I'd have liked to be able to use the trust store configured in the <ssl/truststore> section of the WildFly configuration file (standalone.xml). However, I could not find a way to access the aforementioned configuration imperatively, so had to compromise.

stianst commented:
We've got a separate truststore to WildFly. It's fully accessible from within a provider. Take a look at https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/connections/httpclient/DefaultHttpClientFactory.java#L131

brat000012001 commented:
Thanks for the link and for the comments, I'll take another  look at the trust store SPI in keycloak. One thing though: when a client establishes an SSL connection, WildFly uses the trust store configured through &lt;;ssl/truststore&gt; to send a list of CAs back to the client as well as to validate the client cert. However, I do agree that having multiple places to configure a trust store is somewhat unnecessary. I am thinking that perhaps there isn't really a need for PKIX client cert validation in the x509 cert authentication code since the PKIX cert path validation should have already been done during building of the SSL connection.

stianst commented:
That's true. Undertow should take care of validating the cert. Authenticator in Keycloak should just assume the cert is valid and simply locate the corresponding user.

stianst commented:
So actually you're code doesn't need to use the Truststore SPI. Rather in the future we'll make sure Undertow uses the Truststore SPI (or the other way around) so there's only one place to configure the truststore.

stianst commented:
General approach looks good, but I have a few outstanding comments:

* Certificate validation should be left to Undertow. We don't need to verify certificate twice.
* Currently it can only be used as primary authentication means. Would be nice to allow it as a second factory mechanism, but still require password first. In this case it would be required if a user had certificate configured.
* Missing tests. Prior to merging we need to have tests and at the moment we don't have time to write those ourselves.

brat000012001 commented:
Is x509 second factor authentication something you consider a must have feature before merging?

stianst commented:
No, I'm fine with merging without support for x509 as second factor.

dteleguin commented:
Peter, good job! My two cents: having x509 as the only factor is nevertheless a feature of great value. The major use case for client-cert authentication is smartcards / USB tokens, which are always PIN-protected, thus providing overall 2-factor authentication. Hope this will be included into the next KeyCloak release!

stianst commented:
@brat000012001 could you comment on the following please:

* Certificate validation should be left to Undertow. We don't need to verify certificate twice.
* Missing tests. Prior to merging we need to have tests and at the moment we don't have time to write those ourselves.

brat000012001 commented:
I've merged some of my changes related to X509 cert validation based on our discussion. 
Arquillian integration tests is a bit of a learning curve for me

dteleguin commented:
@stianst , if there's a chance this feature will be included into 2.3.0, I'm ready to help @brat000012001 complete this PR.

brat000012001 commented:
Ok, I've removed x509 path validation and merged the changes, the "only" thing left is to implement the integration tests. Is there an  existing way to enable SSL just for a specific text fixture and to add SSL specific configuration to the wildfly's standalone.xml?

stianst commented:
@pdrozd can someone from your side provide some advice on how to test this?

stianst commented:
@brat000012001 sorry it's taken a while to get back to you on this. The current status is that we won't get this merged for 2.3, but let's aim to get it merged for 2.4. I've asked someone on the QE team to provide you some guidance on how to write tests for this. In summary we have our Arquillian based testsuite that already has support for starting Keycloak with SSL enabled. You would need to also add a truststore to this setup. Not sure how you can test the actual authentication, but I believe you'd have to use httpclient as I doubt selenium will support setting up certificates. With regards to the implementation itself I've not had a chance to look at it in detail yet, but I promise to do this early next week. One concern I have is with regards to the built-in revocation status and other certificate validation features. I would have thought these would be the responsibility of WildFly/Undertow to do. If we can provide you with some help to write the tests would you be able to do that next week? @dteleguin could maybe also help here?

stianst commented:
@dteleguin if you can review and try this out that'd be great

dteleguin commented:
@stianst, @brat000012001, I'm proceeding with testing/reviewing now. I'm ready to help with tests as well.

Some notes:
* Selenium does support client certs, at least with PhantomJS 2.1.1 (exactly our case AFAIK), please check [this SO answer](http://sqa.stackexchange.com/a/18220);
* I agree that certificate validation, OCSP and other stuff is completely of WildFly/Undertow's business; but if absolutely necessary, we could wire a full-featured CA (including OCSP responder) and implement full certificate lifecycle. [XiPKI](https://github.com/xipki/xipki) seems a good candidate, luckily I've been tasked at work to comprehend it;
* don't forget that OCSP won't work until [WFCORE-1705](https://issues.jboss.org/browse/WFCORE-1705) is fixed. The bug seems absolutely trivial. Can we fix it in the KeyCloak distro? Can we somehow encourage WildFly guys to fix it on their side?

brat000012001 commented:
@stianst @dteleguin A couple of things. 
I will be pushing the unit tests pertaining to X509 auth for review sometime today. The unit tests validate the implemented X509 authenticators and OCSP/CRL status checking.
While I do agree that PKIX path validation should be delegated to Undertow given the current state of integration between keycloak and Undertow, I am so sure about OCSP/CRL. I (as a user of the system) wanted to have a level of control over OCSP/CRL checking that Wildfly does not seem to currently provide (I could be wrong of course). Particularly, I wanted to be able to do OCSP/CRL status check exclusively when logging in using X509 client certificates, but not when making requests for resources external to keycloak. One of the requirements was to support retrieving CRL lists from an LDAP directory, a local file or from an external resource using HTTP/HTPS.
Thanks for stepping up to help, @dteleguin!

stianst commented:
Just checked and WildFly doesn't have these features currently anyways so let's keep them in. At some point we'll at least refactor this to have a trustmanager within Keycloak that it can delegate to rather than having these features added directly to the authenticator. That way it can be shared (and also pluggable). We can do that later though.

pdrozd commented:
Few notes about the arquillian tests:
In the arquillian testsuite there is a support for wildfly SSL configuration. Run the testsuite with profile auth-server-wildfly and property auth.server.ssl.required:
`mvn clean install -f testsuite/integration-arquillian/pom.xml -Pauth-server-wildfly -Dauth.server.ssl.required`
This will start the wildfly container with predefined ssl configuration:
Truststore: [testsuite/integration-arquillian/servers/auth-server/jboss/common/keystore/keycloak.truststore](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/servers/auth-server/jboss/common/keystore/keycloak.truststore)
Private key for ssl connection: [testsuite/integration-arquillian/servers/auth-server/jboss/common/keystore/keycloak.jks](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/servers/auth-server/jboss/common/keystore/keycloak.jks)
If you need to modify wildfly configuration (like adding verify-client="requested") change the xsl template [testsuite/integration-arquillian/servers/auth-server/jboss/common/security-wildfly.xsl](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/servers/auth-server/jboss/common/security-wildfly.xsl)
PhantomJS configuration can be added to to arquillian.xml ([testsuite/integration-arquillian/tests/base/src/test/resources/](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/resources/arquillian.xml)

Important note:
If you run the testsuite with the parameter auth.server.ssl.required, most of the tests will fail. This parameter is mostly designed to be used in adapter tests (where the secured applicatin is running on the separate container). The adapter tests are prepared for changing the port and schema, but the tests in the base testsuite aren't. I will prepare fix for the base testsuite soon.

dteleguin commented:
OK, I gave this a try, and all I can say is - good job Peter!

Everything works flawlessly, at least in the simplest scenario I've tried (yet without CRL and OCSP - I plan to test this later). My setup:
* Linux 4.8
* Firefox 49
* Oracle JDK 1.8.0_102 (beware, OpenJDK doesn't support EC cryptography!)
* Aladdin JaCarta PKI (aka Athena/NXP IDProtect USB) with PKCS#11 driver
* XiPKI CA

There is one minor improvement  I can suggest right now. When inferring identity from certificate, extracting from SubjectDN is a special case (and the most widely used one I guess). SubjectDN has a well-known and strict format (a sequence of key-value pairs), so we could simply extract individual keys instead of making the user deal with some obscure regexps. Moreover, I'm not sure if the standard requires that the key-value pairs follow in order, nor am I sure about upper/lower case; trying to craft a universal regexp will only make it more complicated. Instead of "A regular expression to extract user identity" field, I would propose a field that would accept a key name (with a dropdown for most common names like CN, email etc.) This will cover 99% use cases I guess; for the remaining 1%, the field should also accept or a regexp.

I've spent some hour just to figure out what was wrong with my certificate and why it could not be mapped to an existing user (finally did it by inserting a debug log statement). Could we offer some functionality to test our configuration on an existing certificate (either current or uploaded one)? However I'm not sure if this could be done from the authenticator configuration page; @stianst could you please clarify? Can we inject arbitrary markup there?

dteleguin commented:
@pdrozd, chances are we'll have to roll a PKI within integration tests. To me, a good candidate is XiPKI, which is an Apache Karaf application. Could you please suggest how easy/hard it would be to run a Karaf container from within Arquillian tests?

brat000012001 commented:
@dteleguin  Thank you for having taken time to review and for the comments. The motivation for using the regular expressions for extracting user identity was driven by the requirements mandating the user identity encoded in the CN field as a numerical value, i.e. simply extracting CN and using the value was not an option. The functionality can certainly be extended, as you have suggested, to offer a way for users to select from a list of pre-defined patterns (i.e. CN, email, etc.). It should be trivial to add more choices to the existing drop down, i.e. "SubjectDN:email", "SubjectDN:CN" and "SubjectDN". The regular expression can then be applied, if specified, to the extracted value of the DN attribute.

brat000012001 commented:
@pdrozd  Excellent info, appreciate it.

stianst commented:
@pdrozd @dteleguin for testing with PKI we should use an external PKI server. We'll need to be able to support different PKI servers and these should be controlled out side of the testsuite. If we could have one embedded option that is started as part of the tests that would be good as well though.

mposolda commented:
Great stuff! Few notes from me:

- Mockito dependency should be probably removed and replaced with proper integration tests. But I guess you plan to do that anyway?

- IMO it is not good that X509ClientCertificateAuthenticator always show the confirmation page after successful authentication. This behaviour shouldn't be a responsibility of this authenticator IMO.

A while ago, we discussed the possibility about implementing separate "Account Chooser" authenticator, which will be responsible to show the page with confirming identity (like "Do you really want to login as user XY?" ). This may be useful especially for all the similar authenticators without user interaction (Client certs, Kerberos etc). Then admin can always add this Authenticator to the Authentication flow if he wants the "identity confirmation" page to be shown.

- ValidateX509CertificateThumbprint -- I am not sure about this. It seems this should be either removed or enhanced, but in it's current form, it looks unusable as there is not a way to set the certificate thumbrint credential to UserModel. You just set it manually in tests AFAIK. Or am I missing something here?

- It seems the OCSP validation is using Sun internal classes, which is not very good. Besides the https://issues.jboss.org/browse/WFCORE-1705, it won't work on IBM JDK. 
It seems that Bouncycastle has some support for OCSP. Isn't it possible to use Bouncycastle instead? Or we can maybe check what Elytron is doing, as they already have support for OCSP.

- @dteleguin I think that you can tweak the Authenticator page if you add separate angular route to the app.js specific to your authenticator. In that case, it won't use generic page for configure authenticator ( authenticator-config.html ), but will use the page where you can add some nice UI validation helper, which will allow to check if username was parsed correctly from your X509 certificate subject. But I am not 100% sure if we really need that...

The PR is very good start IMO. Maybe in the future we can enhance the functionality even more. For example:

- Keycloak itself will be able to create certificates and sign them with some pre-defined root CA. This will allow us to use something, which we have for OTP (eg. users can setup their client certificates in Account management or after successful login through requiredAction. Once Keycloak creates the client certificate, user will download it and set to his browser)

- Authenticator can optionally provide a possibility to create user if it doesn't yet exists. User will be filled with various info from certificate. AFAIK X509Certificate of user can hold info about various things of user identity (email, firstName and lastName etc). However this will require to use separate mapper SPI to be able to map the attrs from certificate into Keycloak identity...

brat000012001 commented:
@mposolda Thank you, all are great points. 

I do plan on submitting a set of integration tests. 

I agree that the X509 authenticator can be refactored by moving the confirmation page out into a separate "identity confiirmation" authenticator, and given time I am sure it can be done. In the meantime and as a temporary compromise I could enhance the X509 authenticator configuration by adding a check box allowing the users to enable or disable the identity confirmation page,  while the exact design of the "identity confirmation" and how it fits with other authenticators (i.e. Kerberos, etc.) is being ironed out. 

You are correct, without corresponding changes to the user's credentials page, the thumbprint authenticator is not very useful. I can either remove it or, if the changes are simple, make necessary changes to the UI. 

Regarding OCSP: It is possible to use bouncy castle, however re-implementing OCSP check using Bouncy Castle API to mimic the behavior of Sun's implementation is going to take time.

mposolda commented:
> I do plan on submitting a set of integration tests in addition to the mockito unit tests.

The question is, if we shouldn't rather remove mockito entirely? So far, we were trying to avoid it in Keycloak because:
- Unit tests are used just for testing some really simple functionality, which doesn't need to involve HTTP requests interactions etc. For example see JsonParserTest
- Integration tests: Those are all in the integration-arquillian and they test with real server. This is better for real testing than mocks

> I agree that the X509 authenticator can be refactored by moving the confirmation page out into a separate 
> "identity confiirmation" authenticator. In the meantime and as a temporary compromise I could enhance 
> the X509 authenticator configuration by adding a check box allowing the users to enable or disable the 
> identity confirmation page, while the exact design of the "identity confirmation" and how it fits with other 
> authenticators (i.e. Kerberos, etc.) is being ironed out.

+1 . The flag is probably sufficient. We can add "Account selector" authenticator later if needed.

> You are correct, without corresponding changes to the user's credentials page, the thumbprint 
> authenticator is not very useful. I can either remove it or, if the changes are simple, make necessary 
> changes to the UI.

Maybe there should be separate JIRA and PR for that? I don't know if verification of thumbrint is needed at all, but if yes, then it should be likely supported also for browser authenticators, not just for direct grant IMO.

> Regarding OCSP. I am assuming that Sun's OCSP fully implements RFC 6960. Surely, it should be 
> possible to use bouncy castle, however re-implementing OCSP check using Bouncy Castle API to mimic 
> the behavior of Sun's implementation is likely to take time.

If it's possible to workaround https://issues.jboss.org/browse/WFCORE-1705 then we can probably postpone that and just create JIRA. Otherwise we probably need to address that or remove OCSP at all. I don't count that  https://issues.jboss.org/browse/WFCORE-1705 will be fixed anytime soon. And instead of pushing wildfly team to fix it in 10.X and wait for the release etc or do our own tweaks in the distribution, it's probably easier to re-implement OCSP to use bouncycastle :) IBM JDK is probably not a blocker (unless there will be NoClassDefFoundError or something, but that can be probably workarounded by using reflection for dealing with sun specific classes). But still, working IBM JDK will be another bonus point :)

stianst commented:
Mockito tests is not something we want as it doesn't fully test anything and it's a maintenance burden. We're not going to accept a PR that adds Mockito. This is a discussion we've had several times and each time we've come to the conclusion we're not going to use Mock frameworks.

We do need IBM JDK support (sadly) so it should use Bouncycastle. Ideally that's done now, otherwise we'd have to redo it in the future.

For WFCORE-1705 we can probably workaround it by adding an additional module that includes the required paths.

mposolda commented:
If we use bouncycastle, we probably don't need WFCORE-1705 at all AFAIK. Advantage of bouncycastle are 2 things actually (IBM JDK and no need for WFCORE-1705).

brat000012001 commented:
An update: 
- @mposolda ValidateX509CertificateThumbprint authenticator has been removed
- @mposolda  @stianst OCSP validation has been re-implemented to use bouncycastle API; this takes care of WFCORE-1705. 
   - Disclosure: the responder certificate's revocation status is not currently checked, something to look into later on. This functionality is also missing from Sun's OCSP client implementation. 
- mockito tests have been removed from the PR
- @dteleguin Additional identity extractors, in addition to the regular  expression based extractors, to make it easier to extract CN and E-mail attributes from X500Name
- @mposolda @stianst Integration tests to validate X509 authentication and OCSP certificate revocation. I've put together the instructions how to run the integration tests. When running the tests, make sure to start a simple OCSP server using openssl (the instructions how to do it are in the README as well).

zregvart commented:
Hi guys, I hope you don't mind me jumping in. I've made a pull request on @brat000012001 fork with code cleanups and stub OCSP responder so that the integration tests can be run without the external OCSP responder (provided by openssl, as described in the README).

I've asked @brat000012001 to incorporate these changes into this PR.

Furthermore more I would like to investigate the need for custom certificate validation I think that that code can be replaced by using CertPath API. These things are difficult to get right with all the rules outlined in the PKIX RFCs, so it's best to leave that to the underlying JCE implementation.

brat000012001 commented:
@zregvart  Certainly, it may seem like a PITA to run an OCSP using openssl or any other external OCSP responder. I agree that having a CA/OCSP as a part of arquillian tests is a better approach. At some point we discussed a possibility of integrating an existing CA into arquillian tests (see @dteleguin 's comments on Oct 19/Oct 20). OpenSSL approach (IMHO) does provide a way to verify the OCSP validation using an openssl CA/OCSP. While not ideal, it serves as a temporary solution until a full-fledged CA/OCSP server is in place and is running as a part of the integration tests. The advantage of the stub OCSP is that it removes a dependency on openssl in order to run the OCSP integration tests. However, it does not address the need for a full-fledged CA. So neither approach is ideal.
What do you guys think?

brat000012001 commented:
Anyone? Otherwise I am going to go ahead and merge Zoran's changes.

brat000012001 commented:
JIRA KEYCLOAK-4335

stianst commented:
@mposolda we should prioritize this now. Can you review it again and see if there's anything missing or if this can be merged.

LiamMaru commented:
@brat000012001 Thanks for your hard work on this. The organisation I work for are interested using X509 authentication as a second authentication factor; this work is 'just in case' and something may never eventuate, but I'm looking into it nonetheless.

Is there a reason why you didn't implement this as an optional second factor, other than 'didn't need it'? Not to suggest that isn't a perfectly good reason :). I'm just curious incase there is some gotcha that I'll come up against if I do proceed with the modifications required to use X509 as a second factor.

brat000012001 commented:
@LiamMaru The initial requirements didn't ask for the ability to do multi-factor authentication using X.509 certificates, instead relying on smart card PIN code as the first factor. Other than the lack of browser support to programmatically clear out any cached certificates, I do not believe there are gotchas that I am aware of that would prevent having x.509 as a second factor, the devil is in details though :-)

LiamMaru commented:
Cheers @brat000012001, appreciate the reply :). I agree - where browser-based certificate authentication is concerned, the devil is most certainly in the detail. Although the organisation I work for would like to deprecate certificate-based authentication, the word may come down from above that we must continue to support certificates. 

I've done some preliminary work on this and so far and haven't encountered any scary bits (yet). Fingers crossed...

mposolda commented:
I've finally tried to review, sorry it took so long.

PR looks good to me. I don't have any major concerns besides what I already mentioned in previous comment from 15th October 2016  (eg. confirmation screen should be probably moved to separate authenticator, possibility to have auto-registration of new user with successfully authenticated X509 certificate etc). But those are new features or further improvements and can be added later. IMO I would rather merge this PR now to have it ready for Keycloak 3.0.0.CR1, so we can have early feedback from community and we can then add more advanced stuff later.

Few minor comments, which should be fixed in this PR. I hope all of them will be pretty easy to do:
- PR needs to rebase as there are some conflicts now

- README file should be removed from services\src\main\java\org\keycloak\authentication\authenticators\x509\README.md and should be coverted into proper docs. I think that docs PR can be done once this PR is merged. 

- It will be good to add a short note to https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/HOW-TO-RUN.md about how the x509 tests can be executed and how the maven command should look like.

- logging: Logging level of some "info" and "debug" messages should be rather decreased to "debug" and "trace" . Especially the info message in X509ClientCertificateAuthenticator and ValidateX509CertificateUsername for case when certificates are not found in request, should be decreased to debug.

- Exception handling: There is minor issues with same catch block for both GeneralSecurityException and Exception . This can be simplified to one catch block. It is in X509ClientCertificateAuthenticator and ValidateX509CertificateUsername. Also it will be good to at least log an exception stacktrace with DEBUG or WARNING priority to server.log to ensure that exception stacktrace is not "eaten" and can be investigated if needed. On the other hand, displaying exception.getMessage() in the UI is likely not so good as it can teoretically cause some issues (injections, confused users etc).

- There are 79 commits in this PR. I guess it's because you used merge instead of rebase. If it's possible to rather clearify history a bit and cherry-pick the commits on top of latest master, it will be cool. However if it's too much pain, then np :)

Overally, nice work and Thanks a lot for contribution!

brat000012001 commented:
Hey Marek,
The merge conflits've been resolved.  
* Removed README md. I will update the docs once the PR is merged. 
* Added a short note to HOW-TO-RUN.md with instruction how to x.509 related integration tests
* Decreased the logging level where appropriate 
* Merged the exception handlers for GeneralSecurityException and Exception + stacktrace logging

mposolda commented:
@brat000012001 Thanks!

@stianst What do you think? Can we merge or do you have additional feedback?

stianst commented:
@brat000012001 can we get this squashed to a single commit please?

patriot1burke commented:
With just a quick 10 min review, it looks good.  Correct me if I'm wrong, but what seems to be missing is the ability to configure a client cert on the fly for a user.  Would be cool if client certs could be configured on the fly like we do for OTP.  When this PR is merged, this should be added as a JIRA for later consideration.

stianst commented:
What do you mean about certs be configured on the fly? Do you mean a user could be requested to provide the certificate themselves? Wouldn't the more common way certificates work is that an admin sets up the certificate and passes it to the user, rather than the user self-registering it like for OTP?

stianst commented:
@brat000012001  So looks like we're ready with the implementation. Sorry for this taking so long and thanks for your perseverance. Last thing is that we need the PR for documentation to be ready before we merge. Could you send a PR to https://github.com/keycloak/keycloak-documentation for the documentation please? Also, if you could squash to a single commit that'd be great.

mposolda commented:
I also think that optional ability for users to self-create certificates (either through account mgmt or through requiredActions) will be nice. 

Another nice improvements would be:
- Possibility to auto-register authenticated user and prefill his user profile with the info from certificate
- Remove the option for showing the "confirmation page" from the X509Authenticator and rather have separate "account selector" authenticator.

The 2 last things will be useful for similar authenticators like kerberos. But maybe we can wait what community/customers ask before we add new features?

zregvart commented:
I've never found certificate authentication deployed that allows the users to pick their own certificates. Usually you would have security policy in place and established trust with a handful of CA's (trust list), or even enterprise wide deployed CA (common in Microsoft environments). IMHO, the approach of mapping certificates to users, as implemented in this PR, using configuration is the correct one.
It would be a novel idea to include integration with CA so that the users can request certificates from KeyCloak, by simply linking to CA Certificate request page, or using SCEP/ACME/CMP to request one for the user.

brat000012001 commented:
@stianst  

> Could you send a PR to https://github.com/keycloak/keycloak-documentation for the documentation please?
Sure thing, will do.
As for the squashing the changes into a single commit, I will try to get to it in a next couple of days. 
Cheers

brat000012001 commented:
FYI: The continuous integration failed due to an NPE thrown by testsuite\integration-arquillian\tests\base\src\test\java\org\keycloak\testsuite\admin\UserTest.java,esetUserPassword test. I do not believe the build issue was caused by any of my changes

brat000012001 commented:
The documentation PR: https://github.com/keycloak/keycloak-documentation/pull/56 (KEYCLOAK-4610)

stianst commented:
@mposolda can you sanity check the PR now that is a single commit and also review the docs? If it's all OK please merge.

mposolda commented:
@brat000012001 @zregvart @dteleguin Finally merged :) Thanks everyone for your contributions and feedback on this!

stianst commented:
+1 Thanks for contributions and feedback

nelvadas commented:
+1, When is the 3.1.0.Final released planned?
